tsfc/kernel_interface/firedrake_loopy.py

```diff
@@ -0,0 +1,354 @@
+                             for element in coefficient.ufl_element().sub_elements()]
+                    coefficients.extend(split)
+                    self.coefficient_split[coefficient] = split
+                else:
+                    coefficients.append(coefficient)
+                # This is which coefficient in the original form the
+                # current coefficient is.
+                # Consider f*v*dx + g*v*ds, the full form contains two
+                # coefficients, but each integral only requires one.
+                coefficient_numbers.append(form_data.original_coefficient_positions[i])
+        for i, coefficient in enumerate(coefficients):
+            self.coefficient_args.append(
+                self._coefficient(coefficient, "w_%d" % i))
+        self.kernel.coefficient_numbers = tuple(coefficient_numbers)
+
+    def require_cell_orientations(self):
+        """Set that the kernel requires cell orientations."""
+        self.kernel.oriented = True
+
+    def construct_kernel(self, name, impero_c, precision, index_names):
+        """Construct a fully built :class:`Kernel`.
+
+        This function contains the logic for building the argument
+        list for assembly kernels.
+
+        :arg name: function name
+        :arg impero_c: ImperoC tuple with Impero AST and other data
+        :arg precision: floating-point precision for printing
+        :arg index_names: pre-assigned index names
+        :returns: :class:`Kernel` object
+        """
+
+        args = [self.local_tensor, self.coordinates_arg]
+        if self.kernel.oriented:
+            args.append(self.cell_orientations_loopy_arg)
+        args.extend(self.coefficient_args)
+        if self.kernel.integral_type in ["exterior_facet", "exterior_facet_vert"]:
+            args.append(lp.GlobalArg("facet", dtype=numpy.uint32, shape=(1,)))
+        elif self.kernel.integral_type in ["interior_facet", "interior_facet_vert"]:
+            args.append(lp.GlobalArg("facet", dtype=numpy.uint32, shape=(2,)))
+
+        self.kernel.ast = generate_loopy(impero_c, args, precision, name, index_names)
+        return self.kernel
+
+    def construct_empty_kernel(self, name):
+        """Return None, since Firedrake needs no empty kernels.
+
+        :arg name: function name
+        :returns: None
+        """
+        return None
+
+    def get_loopy_arguments(self):
+        args = [self.local_tensor, self.coordinates_arg]
+        if self.kernel.oriented:
+            args.append(self.cell_orientations_loopy_arg)
+        args.extend(self.coefficient_args)
+        if self.kernel.integral_type in ["exterior_facet", "exterior_facet_vert"]:
+            args.append(lp.GlobalArg("facet", dtype=numpy.uint32, shape=(1,)))
+        elif self.kernel.integral_type in ["interior_facet", "interior_facet_vert"]:
+            args.append(lp.GlobalArg("facet", dtype=numpy.uint32, shape=(2,)))
```

Comment:
Duplicate code (the same stuff appears in `construct_kernel`).

Reply:
I've forgotten why I need this function in the first place, deleted.